mlflow/transformers.py

```diff
@@ -0,0 +1,800 @@
+        else model_or_pipeline[_MODEL_KEY]
+    )
+    card_loader = _hub_access()
+    if card_loader:
+        return card_loader.ModelCard.load(model.name_or_path)
+
+
+def _build_pipeline_from_model_input(model, task: str):
+    """
+    Utility for generating a pipeline from component parts. If required components are not
+    specified, use the transformers library pipeline component validation to force raising an
+    exception. The underlying Exception thrown in transformers is verbose enough for diagnosis.
+    """
+    from transformers import pipeline
+
+    pipeline_config = {
+        "task": task,
+        **model,
+    }
+    try:
+        return pipeline(**pipeline_config)
+    except Exception as e:
+        raise MlflowException(
+            "The provided model configuration cannot be created as a Pipeline. "
+            "Please verify that all required and compatible components are "
+            "specified with the correct keys."
+        ) from e
+
+
+def _record_pipeline_components(pipeline) -> Optional[Dict[str, Any]]:
+    """
+    Utility for recording which components are present in either the generated pipeline iff the
+    supplied save object is not a pipeline or the components of the supplied pipeline object.
+    """
+    components_conf = {}
+    components = []
+    if pipeline.feature_extractor:
+        feature_extractor = pipeline.feature_extractor
+        components_conf.update(
+            {_FEATURE_EXTRACTOR_TYPE_KEY: _get_instance_type(feature_extractor, False)}
+        )
+        components.append(_FEATURE_EXTRACTOR_KEY)
+    if pipeline.tokenizer:
+        tokenizer = pipeline.tokenizer
+        components_conf.update({_TOKENIZER_TYPE_KEY: _get_instance_type(tokenizer, False)})
+        components.append(_TOKENIZER_KEY)
+    if pipeline.image_processor:
+        image_processor = pipeline.image_processor
+        components_conf.update(
+            {_IMAGE_PROCESSOR_TYPE_KEY: _get_instance_type(image_processor, False)}
+        )
+        components.append(_IMAGE_PROCESSOR_KEY)
+    if components:
+        components_conf.update({_COMPONENTS_BINARY_KEY: components})
+    return components_conf
+
+
+def _save_components(
+    root_path: pathlib.Path, component_config: Dict[str, Any], pipeline, processor
+):
+    """
+    Saves non-model pipeline components explicitly to a separate directory path for compatibility
+    with certain older model types. In earlier versions of ``transformers``, inferred feature
+    extractor types could be resolved to their correct processor types. This approach ensures
+    compatibility with the expected pipeline configuration argument assignments in later versions
+    of the ``Pipeline`` class.
+    """
+    component_types = component_config["components"]
+    for component_name in component_types:
+        component = getattr(pipeline, component_name)
+        save_path = str(root_path.joinpath(component_name))
+        component.save_pretrained(save_path)
```

Comment:
Does `save_pretrained` not support a Path object?

Reply:
Odd... the linter had complained that `Pipeline.save_pretrained()` required a `str` path. I had assumed that `pathlib.Path` wasn't supported for the library. I just passed the `Path` object to component save and the pipeline save and it works just fine :) pushing these changes now...